librz/il/definitions/bitvector.c

```diff
@@ -0,0 +1,1042 @@
+	int x_msb = rz_il_bv_msb(x);
+	int y_msb = rz_il_bv_lsb(y);
+
+	if (x_msb && y_msb) {
+		return !rz_il_bv_ule(x, y);
+	}
+
+	if (!x_msb && !y_msb) {
+		return rz_il_bv_ule(x, y);
+	}
+
+	// if x_msb set, y_msb unset => x < y
+	// if x_msb unset, y_msb set => x > y
+	// x != y when reaches here
+	return x_msb ? true : false;
+}
+
+/**
+ * Check if x equals to y
+ * \param x RzILBitVector, operand
+ * \param y RzILBitVector, operand
+ * \return ret int, return 1 if x != y, return 0 if x == y
+ */
+RZ_API int rz_il_bv_cmp(RzILBitVector x, RzILBitVector y) {
+	if (x->len != y->len) {
+		return 1;
+	}
+
+	for (int i = 0; i < x->len; ++i) {
+		if (rz_il_bv_get(x, i) != rz_il_bv_get(y, i)) {
+			return 1;
+		}
+	}
+
+	return 0;
+}
+
+/**
+ * Get the length of bitvector
+ * \param bv RzILBitVector
+ * \return len int, length of bitvector
+ */
+RZ_API ut32 rz_il_bv_len(RzILBitVector bv) {
+	return bv->len;
+}
+
+/**
+ * Convert ut32 to `length`-bits bitvector
+ * \param length int, length of bitvector
+ * \param value ut32, the value to convert
+ * \return bv RzILBitVector, pointer to new bitvector
+ */
+RZ_API RzILBitVector rz_il_bv_new_from_ut32(ut32 length, ut32 value) {
+	RzILBitVector bv = rz_il_bv_new0(32);
+	RzILBitVector ret;
+	int type_size = 32;
+
+	ut32 mask = 0x1;
+	for (int i = 0; i < type_size; ++i) {
+		rz_il_bv_set(bv, i, (value & mask) ? true : false);
+		value >>= 1;
+	}
+
+	if (length == type_size) {
+		return bv;
+	}
+
+	if (length < type_size) {
+		// cut
+		ret = rz_il_bv_cut_head(bv, type_size - length);
+		rz_il_bv_free(bv);
+	} else {
+		// prepend
+		ret = rz_il_bv_prepend_zero(bv, length - type_size);
+		rz_il_bv_free(bv);
+	}
+
+	return ret;
+}
+
+/**
+ * Convert ut64 to `length`-bits bitvector
+ * \param length int, length of bitvector
+ * \param value ut64, the value to convert
+ * \return bv RzILBitVector, pointer to new bitvector
+ */
+RZ_API RzILBitVector rz_il_bv_new_from_ut64(ut32 length, ut64 value) {
+	RzILBitVector bv = rz_il_bv_new0(length);
+	RzILBitVector ret;
+	int type_size = 64;
+
+	ut64 mask = 0x1;
+	for (int i = 0; i < type_size; ++i) {
+		rz_il_bv_set(bv, i, (value & mask) ? true : false);
+		value >>= 1;
+	}
+
+	if (length == type_size) {
+		return bv;
+	}
+
+	if (length < type_size) {
+		// cut
+		ret = rz_il_bv_cut_head(bv, type_size - length);
+		rz_il_bv_free(bv);
+	} else {
+		// prepend
+		ret = rz_il_bv_prepend_zero(bv, length - type_size);
+		rz_il_bv_free(bv);
+	}
+
+	return ret;
+}
+
+ut32 rz_il_bv_hash(RzILBitVector x) {
+	ut32 h = 5381;
+	ut32 x_len = x->len;
+
+	if (!x->bits || !x->len) {
+		return h;
+	}
+	for (ut32 i = 0; i < x_len; ++i) {
+		h = (h + (h << 5)) ^ rz_il_bv_get(x, i);
+	}
+	return h;
+}
+
+ut32 rz_il_bv_to_ut32(RzILBitVector x) {
+	ut32 ret = 0;
+	if (x->len > 32) {
+		//		printf("[Warning] Convert to ut32 may loss some bits\n");
+	}
+	for (int i = 0; i < x->len; ++i) {
+		if (rz_il_bv_get(x, i)) {
+			ret += 0x1U << i;
+		}
+	}
+
+	return ret;
+}
+
+/**
```

Comment:
If you export `rz_il_bv_to_ut64()` as an API, please do the same for `rz_il_bv_to_ut32()` as well, might be useful to add also `ut16` and `ut8` versions.